File: agent-app-go/internal/core/answerability_gate.go
package core

import (
	"fmt"
	"regexp"
	"strings"

	agentcontext "github.com/riwonace/agent-app-go/internal/context"
)

type GateAction string

const (
	ActionProceed        GateAction = "PROCEED"
	ActionSearchMore      GateAction = "SEARCH_MORE"
	ActionDecline        GateAction = "DECLINE"
	ActionPartialAnswer  GateAction = "PARTIAL_ANSWER"
)

// GateResult mirrors data class GateResult.
type GateResult struct {
	Passed            bool
	ClaimCoverage     float64
	RequiredClaims    []string
	SupportedClaims   []string
	MissingClaims     []string
	RecommendedAction GateAction
	Reasoning         string
}

var (
	entityParticleRe = regexp.MustCompile(`[\x{AC00}-\x{D7A3}]{2,}(의|은|는|이|가|을|를)`)
	entitySingleQuote = regexp.MustCompile(`'[^']+'`)
	entityDoubleQuote = regexp.MustCompile(`"[^"]+"`)
	entityStripRe     = regexp.MustCompile(`[의은는이가을를'"]`)

	timeInfoRe     = regexp.MustCompile(`\d{4}|\d+월|\d+일|\d+년`)
	locationInfoRe = regexp.MustCompile(`[\x{AC00}-\x{D7A3}]+(시|도|구|동|로|길)`)
	quantityRe     = regexp.MustCompile(`\d+`)
	personInfoRe1  = regexp.MustCompile(`[\x{AC00}-\x{D7A3}]{2,4}\s*(씨|님|대표|사장|이사|팀장|과장|부장|매니저)`)
	personInfoRe2  = regexp.MustCompile(`담당자|개발자|관리자|작성자`)
)

// AnswerabilityGate ports AnswerabilityGate.kt 1:1: heuristic claim-coverage
// verification. The `useLlm` config flag exists in the baseline but is never actually
// read inside verify() there either -- preserved here as a documented no-op field to
// match that exact (if surprising) baseline behavior, not as a bug to fix.
type AnswerabilityGate struct {
	CoverageThreshold float64
	UseLLM            bool
}

func (g *AnswerabilityGate) Verify(question string, evidence []agentcontext.ContextItem, profile QueryProfile) GateResult {
	requiredClaims := extractRequiredClaims(question, profile)
	supportedClaims := extractSupportedClaims(evidence, requiredClaims)

	var coverage float64
	if len(requiredClaims) == 0 {
		coverage = 1.0
	} else {
		coverage = float64(len(supportedClaims)) / float64(len(requiredClaims))
	}

	supportedSet := make(map[string]bool, len(supportedClaims))
	for _, c := range supportedClaims {
		supportedSet[c] = true
	}
	var missingClaims []string
	for _, c := range requiredClaims {
		if !supportedSet[c] {
			missingClaims = append(missingClaims, c)
		}
	}

	action := determineGateAction(coverage, missingClaims, profile, g.CoverageThreshold)
	passed := action == ActionProceed || action == ActionPartialAnswer

	return GateResult{
		Passed:            passed,
		ClaimCoverage:     coverage,
		RequiredClaims:    requiredClaims,
		SupportedClaims:   supportedClaims,
		MissingClaims:     missingClaims,
		RecommendedAction: action,
		Reasoning:         buildGateReasoning(coverage, len(requiredClaims), len(supportedClaims), action),
	}
}

func extractRequiredClaims(question string, profile QueryProfile) []string {
	var claims []string
	seen := make(map[string]bool)
	add := func(c string) {
		if !seen[c] {
			seen[c] = true
			claims = append(claims, c)
		}
	}

	q := strings.ToLower(question)
	if strings.Contains(q, "누가") || strings.Contains(q, "누구") {
		add("person_identity")
	}
	if strings.Contains(q, "언제") {
		add("time_info")
	}
	if strings.Contains(q, "어디") || strings.Contains(q, "위치") {
		add("location_info")
	}
	if strings.Contains(q, "왜") || strings.Contains(q, "이유") {
		add("reason_explanation")
	}
	if strings.Contains(q, "어떻게") || strings.Contains(q, "방법") {
		add("method_process")
	}
	if strings.Contains(q, "얼마") || strings.Contains(q, "몇") {
		add("quantity_value")
	}

	switch profile.Intent {
	case IntentFactual:
		add("factual_data")
	case IntentAggregation:
		add("aggregated_value")
	case IntentComparison:
		add("comparison_item_a")
		add("comparison_item_b")
	case IntentExplanation:
		add("concept_definition")
	case IntentMultiHop:
		add("intermediate_result")
		add("final_result")
	case IntentRelation:
		add("relationship_info")
	}

	for _, pattern := range []*regexp.Regexp{entityParticleRe, entitySingleQuote, entityDoubleQuote} {
		for _, match := range pattern.FindAllString(question, -1) {
			entity := strings.TrimSpace(entityStripRe.ReplaceAllString(match, ""))
			if len([]rune(entity)) >= 2 {
				add("entity_" + entity)
			}
		}
	}

	return claims
}

func extractSupportedClaims(evidence []agentcontext.ContextItem, requiredClaims []string) []string {
	var texts []string
	hasSQLSource := false
	hasGraphSource := false
	for _, e := range evidence {
		texts = append(texts, e.Text)
		if e.Source == "sql" {
			hasSQLSource = true
		}
		if e.Source == "knowledge-graph" {
			hasGraphSource = true
		}
	}
	combinedText := strings.ToLower(strings.Join(texts, " "))

	var supported []string
	for _, claim := range requiredClaims {
		switch {
		case claim == "person_identity" && anyHasPersonInfo(evidence):
			supported = append(supported, claim)
		case claim == "time_info" && timeInfoRe.MatchString(combinedText):
			supported = append(supported, claim)
		case claim == "location_info" && locationInfoRe.MatchString(combinedText):
			supported = append(supported, claim)
		case claim == "quantity_value" && quantityRe.MatchString(combinedText):
			supported = append(supported, claim)
		case claim == "factual_data" && len(evidence) > 0:
			supported = append(supported, claim)
		case claim == "aggregated_value" && hasSQLSource:
			supported = append(supported, claim)
		case strings.HasPrefix(claim, "comparison_item_") && len(evidence) >= 2:
			supported = append(supported, claim)
		case claim == "concept_definition" && len([]rune(combinedText)) > 100:
			supported = append(supported, claim)
		case claim == "relationship_info" && hasGraphSource:
			supported = append(supported, claim)
		case strings.HasPrefix(claim, "entity_"):
			entityName := strings.TrimPrefix(claim, "entity_")
			if strings.Contains(combinedText, entityName) {
				supported = append(supported, claim)
			}
		case claim == "intermediate_result" && len(evidence) >= 1:
			supported = append(supported, claim)
		case claim == "final_result" && len(evidence) > 0:
			supported = append(supported, claim)
		case claim == "reason_explanation" && len([]rune(combinedText)) > 50:
			supported = append(supported, claim)
		case claim == "method_process" && len([]rune(combinedText)) > 50:
			supported = append(supported, claim)
		}
	}

	return supported
}

func anyHasPersonInfo(evidence []agentcontext.ContextItem) bool {
	for _, e := range evidence {
		if hasPersonInfo(e.Text) {
			return true
		}
	}
	return false
}

func hasPersonInfo(text string) bool {
	return personInfoRe1.MatchString(text) || personInfoRe2.MatchString(text)
}

func determineGateAction(coverage float64, missingClaims []string, profile QueryProfile, threshold float64) GateAction {
	switch {
	case coverage >= threshold:
		return ActionProceed
	case coverage >= 0.5:
		for _, c := range missingClaims {
			if c == "factual_data" || c == "aggregated_value" || c == "quantity_value" {
				return ActionSearchMore
			}
		}
		return ActionPartialAnswer
	case coverage >= 0.3:
		if profile.Intent == IntentExplanation {
			return ActionPartialAnswer
		}
		return ActionSearchMore
	default:
		return ActionDecline
	}
}

func buildGateReasoning(coverage float64, required, supported int, action GateAction) string {
	return fmt.Sprintf("커버리지 %.0f%% (%d/%d 클레임) → %s", coverage*100, supported, required, action)
}

File: agent-app-ktor/src/main/kotlin/com/riwonace/agentktor/core/AnswerabilityGate.kt
package com.riwonace.agentktor.core

import com.riwonace.agentktor.context.ContextItem
import org.slf4j.LoggerFactory

/**
 * Answerability Gate: 수집된 증거가 질문에 답할 수 있는지 검증한다.
 *
 * 핵심 메트릭: claim_coverage = supported_claims / required_claims
 *
 * claim_coverage < 임계값이면:
 * - 추가 검색 요청
 * - 또는 "답변 불가" 응답 생성
 *
 * baseline's constructor also takes a Spring AI ChatClient and a useLlm flag; neither is
 * ever read inside verify() there (an existing dead field/flag in the Kotlin/Spring AI
 * original, not something this port introduces) -- since there's no equivalent
 * ChatClient type outside Spring AI, both are dropped here rather than carried as an
 * unused Any? placeholder. Behavior is identical either way: useLlm changes nothing.
 */
class AnswerabilityGate(
    private val coverageThreshold: Double = 0.7,
    private val useLlm: Boolean = false,
) {
    private val log = LoggerFactory.getLogger(javaClass)

    /**
     * 검증 결과
     */
    data class GateResult(
        /** 통과 여부 */
        val passed: Boolean,
        /** 클레임 커버리지 (0.0 ~ 1.0) */
        val claimCoverage: Double,
        /** 감지된 필수 클레임들 */
        val requiredClaims: List<String>,
        /** 지원되는 클레임들 */
        val supportedClaims: List<String>,
        /** 누락된 클레임들 */
        val missingClaims: List<String>,
        /** 권장 액션 */
        val recommendedAction: GateAction,
        /** 검증 사유 */
        val reasoning: String,
    )

    enum class GateAction {
        /** 답변 생성 진행 */
        PROCEED,
        /** 추가 증거 검색 필요 */
        SEARCH_MORE,
        /** 답변 불가 응답 생성 */
        DECLINE,
        /** 부분 답변 (가능한 것만) */
        PARTIAL_ANSWER,
    }

    /**
     * 증거가 질문에 답하기 충분한지 검증한다.
     */
    fun verify(
        question: String,
        evidence: List<ContextItem>,
        profile: QueryProfile,
    ): GateResult {
        // 1. 질문에서 필수 클레임 추출
        val requiredClaims = extractRequiredClaims(question, profile)

        // 2. 증거에서 지원 가능한 클레임 추출
        val supportedClaims = extractSupportedClaims(evidence, requiredClaims)

        // 3. 커버리지 계산
        val coverage = if (requiredClaims.isEmpty()) {
            1.0
        } else {
            supportedClaims.size.toDouble() / requiredClaims.size
        }

        // 4. 누락 클레임 식별
        val missingClaims = requiredClaims - supportedClaims.toSet()

        // 5. 액션 결정
        val action = determineAction(coverage, missingClaims, profile)

        val passed = action == GateAction.PROCEED || action == GateAction.PARTIAL_ANSWER

        log.info(
            "Answerability 검증: coverage={}, required={}, supported={}, missing={}, action={}",
            "%.2f".format(coverage), requiredClaims.size, supportedClaims.size, missingClaims.size, action,
        )

        return GateResult(
            passed = passed,
            claimCoverage = coverage,
            requiredClaims = requiredClaims,
            supportedClaims = supportedClaims,
            missingClaims = missingClaims,
            recommendedAction = action,
            reasoning = buildReasoning(coverage, requiredClaims.size, supportedClaims.size, action),
        )
    }

    /**
     * 질문에서 필수 클레임 추출 (휴리스틱 기반)
     */
    private fun extractRequiredClaims(question: String, profile: QueryProfile): List<String> {
        val claims = mutableListOf<String>()
        val q = question.lowercase()

        // 의문사 기반 클레임
        if (q.contains("누가") || q.contains("누구")) claims += "person_identity"
        if (q.contains("언제")) claims += "time_info"
        if (q.contains("어디") || q.contains("위치")) claims += "location_info"
        if (q.contains("왜") || q.contains("이유")) claims += "reason_explanation"
        if (q.contains("어떻게") || q.contains("방법")) claims += "method_process"
        if (q.contains("얼마") || q.contains("몇")) claims += "quantity_value"

        // 의도별 필수 클레임
        when (profile.intent) {
            QueryIntent.FACTUAL -> claims += "factual_data"
            QueryIntent.AGGREGATION -> claims += "aggregated_value"
            QueryIntent.COMPARISON -> {
                claims += "comparison_item_a"
                claims += "comparison_item_b"
            }
            QueryIntent.EXPLANATION -> claims += "concept_definition"
            QueryIntent.MULTI_HOP -> {
                claims += "intermediate_result"
                claims += "final_result"
            }
            QueryIntent.RELATION -> claims += "relationship_info"
        }

        // 엔티티 추출 (간단한 패턴)
        val entityPatterns = listOf(
            Regex("[가-힣]{2,}(의|은|는|이|가|을|를)"),
            Regex("'[^']+'"),
            Regex("\"[^\"]+\""),
        )
        for (pattern in entityPatterns) {
            pattern.findAll(question).forEach { match ->
                val entity = match.value.replace(Regex("[의은는이가을를'\"]"), "").trim()
                if (entity.length >= 2) {
                    claims += "entity_$entity"
                }
            }
        }

        return claims.distinct()
    }

    /**
     * 증거에서 지원 가능한 클레임 추출
     */
    private fun extractSupportedClaims(
        evidence: List<ContextItem>,
        requiredClaims: List<String>,
    ): List<String> {
        val supported = mutableListOf<String>()
        val combinedText = evidence.joinToString(" ") { it.text }.lowercase()

        for (claim in requiredClaims) {
            when {
                claim == "person_identity" && evidence.any { hasPersonInfo(it.text) } ->
                    supported += claim

                claim == "time_info" && combinedText.contains(Regex("\\d{4}|\\d+월|\\d+일|\\d+년")) ->
                    supported += claim

                claim == "location_info" && combinedText.contains(Regex("[가-힣]+(시|도|구|동|로|길)")) ->
                    supported += claim

                claim == "quantity_value" && combinedText.contains(Regex("\\d+")) ->
                    supported += claim

                claim == "factual_data" && evidence.isNotEmpty() ->
                    supported += claim

                claim == "aggregated_value" && evidence.any { it.source == "sql" } ->
                    supported += claim

                claim.startsWith("comparison_item_") && evidence.size >= 2 ->
                    supported += claim

                claim == "concept_definition" && combinedText.length > 100 ->
                    supported += claim

                claim == "relationship_info" && evidence.any { it.source == "knowledge-graph" } ->
                    supported += claim

                claim.startsWith("entity_") -> {
                    val entityName = claim.removePrefix("entity_")
                    if (combinedText.contains(entityName)) {
                        supported += claim
                    }
                }

                claim == "intermediate_result" && evidence.size >= 1 ->
                    supported += claim

                claim == "final_result" && evidence.isNotEmpty() ->
                    supported += claim

                claim == "reason_explanation" && combinedText.length > 50 ->
                    supported += claim

                claim == "method_process" && combinedText.length > 50 ->
                    supported += claim
            }
        }

        return supported
    }

    /**
     * 텍스트에 사람 정보가 포함되어 있는지 확인
     */
    private fun hasPersonInfo(text: String): Boolean {
        val personPatterns = listOf(
            Regex("[가-힣]{2,4}\\s*(씨|님|대표|사장|이사|팀장|과장|부장|매니저)"),
            Regex("담당자|개발자|관리자|작성자"),
        )
        return personPatterns.any { it.containsMatchIn(text) }
    }

    /**
     * 커버리지와 상황에 따른 액션 결정
     */
    private fun determineAction(
        coverage: Double,
        missingClaims: List<String>,
        profile: QueryProfile,
    ): GateAction {
        return when {
            // 완전 커버리지
            coverage >= coverageThreshold -> GateAction.PROCEED

            // 부분 커버리지 (50% 이상)
            coverage >= 0.5 -> {
                // 핵심 클레임이 누락되었으면 추가 검색
                val criticalMissing = missingClaims.any {
                    it in listOf("factual_data", "aggregated_value", "quantity_value")
                }
                if (criticalMissing) GateAction.SEARCH_MORE else GateAction.PARTIAL_ANSWER
            }

            // 낮은 커버리지
            coverage >= 0.3 -> {
                // 설명형 질문은 부분 답변 허용
                if (profile.intent == QueryIntent.EXPLANATION) {
                    GateAction.PARTIAL_ANSWER
                } else {
                    GateAction.SEARCH_MORE
                }
            }

            // 매우 낮은 커버리지
            else -> GateAction.DECLINE
        }
    }

    private fun buildReasoning(
        coverage: Double,
        required: Int,
        supported: Int,
        action: GateAction,
    ): String {
        val coveragePercent = String.format("%.0f", coverage * 100)
        return "커버리지 $coveragePercent% ($supported/$required 클레임) → $action"
    }
}
